src/blr/cli/tagfastq.py

```diff
@@ -119,22 +101,42 @@ def parse_corrected_barcodes(open_file):
     return corrected_barcodes
 
 
-def parse_raw_barcodes(open_file):
-    """
-    Parse FASTA/FASTQ containing barcode sequences and return a dictionary with entry headers
-    pointing to a raw barcodes sequence.
-    :param open_file: dnaio odject.
-    :return: dict: entry headers pointing to a raw barcodes sequence
-    """
-    for barcode in tqdm(open_file, desc="Raw barcodes processed"):
-        header, _ = barcode.name.split(maxsplit=1)
-        yield {header: barcode.sequence}
+class BarcodeReader:
+    def __init__(self, filename):
+        self.cache = dict()
+        self.filename = filename
+        self.barcodes = self.parse()
+
+    def parse(self):
+        for barcode in tqdm(self.open_file, desc="Uncorrected barcodes processed"):
+            header, _ = barcode.name.split(maxsplit=1)
+            yield {header: barcode.sequence}
+
+    def get_barcode(self, header, maxiter=10):
+        if header in self.cache:
+            return self.cache.pop(header)
+
+        for header_barcode_pair in islice(self.barcodes, maxiter):
+            # If header in next pair then parser lines are synced --> drop cache.
+            if header in header_barcode_pair:
+                self.cache.clear()
+                return header_barcode_pair[header]
+
+            self.cache.update(header_barcode_pair)
+        return None
+
+    def __enter__(self):
+        self.open_file = dnaio.open(self.filename, mode="r")
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return self.open_file.close()
```

Comment:
As mentioned above, it would be nicer to implement a separate `close()` method. Then you can defer to that for closing. So this could then be replaced with just `self.close()`.
Since `close()` doesn’t return anything, you should also omit the `return`.